### ISSUE TITLE
Proposal: Use a series of chained `Task` functions to coordinate asynchronous steps which can fail

### ISSUE THREAD

**Issue by julianpistorius** [exosphere-project/exosphere#63]
... like server launching. Multi-launch could be done concurrently by using a `Process` for each series of composed tasks. I think this is the _Elm Way ™_ to do _pipelines_ or _ROP_.

See:
- http://package.elm-lang.org/packages/elm-lang/core/latest/Task
- https://web.archive.org/web/20160719013553/http://guide.elm-lang.org:80/error_handling/task.html
- http://package.elm-lang.org/packages/elm-lang/core/latest/Process

I also found a couple of interesting state transition packages which might be useful:

- http://package.elm-lang.org/packages/folkertdev/elm-state/latest
- http://package.elm-lang.org/packages/athanclark/elm-discrete-transition/latest

**Comment by c-mart** [comment]
@julianpistorius I hear you, but we also need to "gracefully recover" if the SPA is terminated unexpectedly. If I start launching a server and then close Exosphere, I would expect things to finish when I logged in again. So, I'd generally prefer to determine "what remains to be done" by querying OpenStack (e.g. "Oh, here's a new instance here but it needs a floating IP, let's fix that now"), than by relying on Exosphere's working memory (unless that is persisted to storage somehow).

What's wrong with the way we're doing multi-launch now (it sounds like `Cmd.batch` does things [mostly in parallel](https://stackoverflow.com/questions/39052337/run-two-commands-at-the-same-time-in-elm))? Or, can you provide another example of what we should be using `Process` for? Otherwise, I'd close this issue as a solution in search of a problem.

**Comment by julianpistorius** [comment]
I agree with you about the recovery argument. I was imagining some combination of these two approaches. 

By having the sequence of steps explicit, in one place, it makes the flow easier to understand and modify.

This was an interesting thread comparing the different approaches:

https://groups.google.com/forum/#!msg/elm-dev/MEBzD3f7Bq8/Mc9rPrjFBgAJ

`Process` might be a red-herring here - I was mainly interested in exploring composing `Tasks` into chains. I don't think there is any problem with the current parallel multi-launch facility. `Process` might have to come into play if we went to a chain of `Tasks` though - not sure.

Anyhoo, I'm not really sure about the various pro's and con's yet, and the current solution works, so no need to rush into anything.

**Comment by lenards** [comment]
So I've run into needing to fetch something _"andThen"_ do something with the results of the first `Task`. I had a devil of a time wrapping my head about that. On Elm Slack, @cmditch was amazingly helpful and provided a nice `retry` example that uses `Process.sleep` to _wait_ between attempts. I think you both might find that example useful. 

Example: [retry](https://github.com/cmditch/elm-ethereum/blob/43eaa4c6fad0e3c13504738cbb886f23601f7590/src/Web3.elm#L226-L246)

And, I recently created an interface where all the "actions" needed to be captured as events, then applied, in order, via calls to corresponding endpoints. Another thing I had issues finding, but I now have done it, so I can work up a _"shareable"_ version of that. During the time I was wrapping my head about it, I found this article, ["Http in Elm"](https://medium.com/@pwentz/http-in-elm-23844fa0e315) written by [pwentz](https://medium.com/@pwentz) on Elm Slack. 

The other thing that I realized is that `Task.map2` has some wonderful applications if you're forced into a _"fetch a list of things, then a list based on that list"_. I imagine that might be something you read into with a cloud provider API.

**Comment by julianpistorius** [comment]
Oh yes - that's exactly what I was hoping for. Thanks @lenards! 🙇
